routing/index_graph.cpp

```diff
@@ -0,0 +1,197 @@
+#include "index_graph.hpp"
+
+#include "base/assert.hpp"
+
+namespace routing
+{
+IndexGraph::IndexGraph(unique_ptr<Geometry> geometry) : m_geometry(move(geometry)) {}
+
+void IndexGraph::GetOutgoingEdgesList(JointId jointId, vector<TEdgeType> & edges) const
+{
+  GetEdgesList(jointId, edges, true);
+}
+
+void IndexGraph::GetIngoingEdgesList(JointId jointId, vector<TEdgeType> & edges) const
+{
+  GetEdgesList(jointId, edges, false);
+}
+
+double IndexGraph::HeuristicCostEstimate(JointId jointFrom, JointId jointTo) const
+{
+  return m_geometry->CalcHeuristic(GetFSeg(jointFrom), GetFSeg(jointTo));
+}
+
+m2::PointD const & IndexGraph::GetPoint(JointId jointId) const
+{
+  return m_geometry->GetPoint(GetFSeg(jointId));
+}
+
+void IndexGraph::Export(vector<Joint> const & joints)
+{
+  m_fsegIndex.Export(joints);
+  BuildJoints(joints.size());
+}
+
+JointId IndexGraph::InsertJoint(FSegId fseg)
+{
+  JointId const existId = m_fsegIndex.GetJointId(fseg);
+  if (existId != kInvalidJointId)
+    return existId;
+
+  JointId const jointId = m_jointOffsets.size();
+  m_jointOffsets.emplace_back(JointOffset(m_fsegs.size(), m_fsegs.size() + 1));
+  m_fsegs.emplace_back(fseg);
+  m_fsegIndex.AddJoint(fseg, jointId);
+  return jointId;
+}
+
+vector<FSegId> IndexGraph::RedressRoute(vector<JointId> const & route) const
+{
+  vector<FSegId> fsegs;
+  if (route.size() < 2)
+  {
+    if (route.size() == 1)
+      fsegs.emplace_back(GetFSeg(route[0]));
+    return fsegs;
+  }
+
+  fsegs.reserve(route.size() * 2);
+
+  for (size_t i = 0; i < route.size() - 1; ++i)
+  {
+    JointId const prevJoint = route[i];
+    JointId const nextJoint = route[i + 1];
+
+    auto const & pair = FindSharedFeature(prevJoint, nextJoint);
+    if (!pair.first.IsValid())
+    {
+      fsegs.clear();
+      return fsegs;
+    }
+
+    if (i == 0)
+      fsegs.push_back(pair.first);
+
+    uint32_t const featureId = pair.first.GetFeatureId();
+    uint32_t const segFrom = pair.first.GetSegId();
+    uint32_t const segTo = pair.second.GetSegId();
+
+    ASSERT_NOT_EQUAL(segFrom, segTo, ());
+
+    if (segFrom < segTo)
+    {
+      for (uint32_t seg = segFrom + 1; seg < segTo; ++seg)
+        fsegs.push_back({featureId, seg});
+    }
+    else
+    {
+      for (uint32_t seg = segFrom - 1; seg > segTo; --seg)
```

Comment:
segFrom can't be 0, because it else branch, so segFrom > segTo (it can't be equal, ASSERT_NOT_EQALS)

segTo can be 0, it is ok.

This code is correct, but really some confusing, so need bit rework to be more clear.